### PR TITLE
Add get_storage_deposit_return_outputs()

### LINKED ISSUE #1082: [Task]: Check unlock conditions in try_select_inputs
### Task description

If there is are expiration/timelock or storage deposit return unlock conditions, we might have to add an output for the storage deposit return, if the output wasn't expired, or we have to at least check that there is an output sending the required amount to the return address

### Requirements

-

### Acceptance criteria

-

### Creation checklist

- [ ] I have assigned this task to the correct people
- [X] I have added the most appropriate labels
- [X] I have linked the correct milestone and/or project

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Add get_storage_deposit_return_outputs(), because this wasn't handled before
Add sdr amounts to the required amounts, so we select enough inputs
Also changed the dedup of inputs, because we could get the same output with multiple addresses

## Links to any relevant issues

Fixes #1082 

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by samuel-rufi — DISMISSED** [review]

File: src/api/block_builder/input_selection/remainder.rs [hunk 1]
```diff
@@ -22,6 +24,74 @@ use crate::{
...
+pub(crate) fn get_storage_deposit_return_outputs<'a>(
+    inputs: impl Iterator<Item = &'a InputSigningData> + Clone,
+    outputs: impl Iterator<Item = &'a Output> + Clone,
+    current_time: u32,
+) -> Result<Vec<Output>> {
+    log::debug!("[get_storage_deposit_return_outputs]");
+
+    // Get inputs with storage deposit return unlock condition
+    let inputs_with_sdr = inputs
+        .filter(|i| i.output.unlock_conditions().unwrap().storage_deposit_return().is_some())
```
Comment (thibault-martinez) [inline src/api/block_builder/input_selection/remainder.rs]: Replace the `unwrap` with `and_then` to be safe.

File: src/api/block_builder/input_selection/remainder.rs [hunk 2]
```diff
@@ -22,6 +24,74 @@ use crate::{
...
+        .map(|i| i.output.clone())
```
Comment (thibault-martinez) [inline src/api/block_builder/input_selection/remainder.rs]: No need to clone

File: src/api/block_builder/input_selection/remainder.rs [hunk 3]
```diff
@@ -22,6 +24,74 @@ use crate::{
...
+        .collect::<Vec<Output>>();
```
Comment (thibault-martinez) [inline src/api/block_builder/input_selection/remainder.rs]: No need to collect

File: src/api/block_builder/input_selection/remainder.rs [hunk 4]
```diff
@@ -22,6 +24,74 @@ use crate::{
...
+
+    // There could be multiple sdr outputs required for the same address, so we keep track of the required amount here.
+    let mut required_address_returns: HashMap<Address, u64> = HashMap::new();
+
+    // Exclude expired outputs
+    for input in inputs_with_sdr {
+        if let Some(sdr) = sdr_not_expired(&input, current_time) {
```
Comment (thibault-martinez) [inline src/api/block_builder/input_selection/remainder.rs]: Maybe you can even use `sdr_not_expired` in the filter instead ?

File: src/api/block_builder/input_selection/remainder.rs [hunk 5]
```diff
@@ -22,6 +24,74 @@ use crate::{
...
+                        false
+                    }
+                } else {
+                    false
+                }
+            })
+            .map(|o| o.amount())
+            .sum();
+
+        if let Ordering::Greater = required_return_amount.cmp(&existing_sdr_amount) {
```
Comment (thibault-martinez) [inline src/api/block_builder/input_selection/remainder.rs]: Using > would be more readable IMO

File: src/api/block_builder/input_selection/remainder.rs [hunk 6]
```diff
@@ -22,6 +24,76 @@ use crate::{
...
+    outputs: impl Iterator<Item = &'a Output> + Clone,
+    current_time: u32,
+) -> Result<Vec<Output>> {
+    log::debug!("[get_storage_deposit_return_outputs]");
+
+    // Get inputs with storage deposit return unlock condition that are not expired
+    let inputs_sdr = inputs
+        .filter(|i| sdr_not_expired(&i.output, current_time).is_some())
+        // Safe to unwrap since we only return outputs with a sdr when filtering with sdr_not_expired()
+        .map(|i| {
```
Comment (thibault-martinez) [inline src/api/block_builder/input_selection/remainder.rs]: You can clean all this using https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.filter_map

**Review by thibault-martinez — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]

**Pull request merged**